src/Logic.hs

```diff
@@ -471,31 +499,42 @@ tryIntegratePullRequest pr state =
 -- target branch. If the push fails, restarts the integration cycle for the
 -- candidate.
 -- TODO: Get rid of the tuple; just pass the ID and do the lookup with fromJust.
-pushCandidate :: (PullRequestId, PullRequest) -> ProjectState -> Action ProjectState
-pushCandidate (pullRequestId, pullRequest) state = do
+pushCandidate :: (PullRequestId, PullRequest) -> Sha -> ProjectState -> Action ProjectState
+pushCandidate (pullRequestId, pullRequest) newHead state =
   -- Look up the sha that will be pushed to the target branch. Also assert that
-  -- the pull request has really been approved and built successfully. If it was
+  -- the pull request has really been approved. If it was
```

Comment:
Ah, wait. So it's no longer doing the check here because the call site already does it?

Reply:
Yeah, call site has already been doing that. So the previous flow looked something like this:
```haskell
  case Pr.integrationStatus pullRequest of
-- ...
    Integrated _sha_we_need buildStatus -> case buildStatus of
      BuildSucceeded -> case Pr.integrationStatus pullRequest of -- pushCandidate
        somethingElseThanBuildSucceeded -> error "This is bad!"
```

This might have made sense if `pushCandidate` was exported from the module, but it isn't.